bson/marshal.go

```diff
@@ -52,14 +52,14 @@ func MarshalAppend(dst []byte, val interface{}) ([]byte, error) {
 // MarshalWithRegistry returns the BSON encoding of val as a BSON document. If val is not a type that can be transformed
 // into a document, MarshalValueWithRegistry should be used instead.
 func MarshalWithRegistry(r *bsoncodec.Registry, val interface{}) ([]byte, error) {
-	dst := make([]byte, 0, 256) // TODO: make the default cap a constant
+	dst := make([]byte, 0)
```

Comment:
Is this just bad practice or is there a performance reason for why we don't want to do this?

Reply:
That is bad practice in the sense that "blind allocations" (i.e. allocations where you're just guessing how much memory you'll need) are an indication that you may be able to reduce allocations by reusing buffers. In this case, it seems allocating the `dst` byte slice with a capacity of 256 is to attempt to reduce the number of allocations when appending a BSON message to `dst` one field at a time. However, that blind allocation is now unnecessary because `MarshalAppendWithContext` only writes to `dst` in `bson.valueWriter#Flush` when the full message length is available, so there will only ever be one allocation for `dst`.


Example old process for encoding a `map`. Note that `vw.buf == dst` in the current code:
1. `dst := make([]byte, 0, 256)` - Allocate 256 bytes for `dst`.
2. Append each key and value encoded as BSON to `dst`.
  a. `vw.buf = bsoncore.AppendHeader(vw.buf, t, key)` - Append field header to `dst`. Possibly allocate-and-copy to grow.
  b. `vw.buf = bsoncore.AppendString(vw.buf, s)` - Append string to `dst`. Possibly allocate-and-copy to grow.
  c. Repeat for other fields...
3. `vw.WriteDocumentEnd()` - Append document end to `dst`. Possibly allocate-and-copy to grow.


Example new process for encoding a `map`. Assume `vw` is being reused and already has a buffer that has been grown to a sufficient capacity from encoding previous messages:
1. `dst := make([]byte, 0)` - Declare empty byte slice `dst`. No allocations.
2. Append each key and value encoded as BSON to `vw.buf`.
  a. `vw.buf = bsoncore.AppendHeader(vw.buf, t, key)` - Append field header to `vw.buf`, which has enough capacity. No allocations.
  b. `vw.buf = bsoncore.AppendString(vw.buf, s)` - Append string to `vw.buf`, which has enough capacity. No allocations.
  c. Repeat for other fields...
3. `vw.WriteDocumentEnd()` - Append document end to `vw.buf` which has enough capacity. Append to `dst` the exact size of `vw.buf`. One allocation.

Reply:
Ah thanks for the explanation here; very helpful!